transformations/contextual_meaning_perturbation/transformation.py

```diff
@@ -0,0 +1,172 @@
+        )
+        if self.verbose:
+            print("Completed loading the cross-lingual XLM-R (base) model.\n")
+
+        if self.verbose:
+            print(
+                "Starting to load Spacy model ("
+                + language
+                + ") for retrieving linguistic features.\n"
+            )
+        if language == "en":
+            self.linguistic_pipeline = spacy.load("en_core_web_sm")
+        elif language == "de":
+            try:
+                self.linguistic_pipeline = spacy.load("de_core_news_sm")
+            except ImportError:
+                print(
+                    "To perturb German text, please download SpaCy's German pipeline de_core_news_sm, \
+            for example by using the following command: python -m spacy download de_core_news_sm"
+                )
+        else:
+            raise NotImplementedError(
+                "As of now, only English and German are supported."
+            )
+        if self.verbose:
+            print("Completed loading Spacy model.\n")
+
+        self.pos_to_change = pos_to_change
+        assert perturbation_rate <= 1 and perturbation_rate >= 0
+        self.percentage = perturbation_rate  # How many of the elligable POS tags should be changed? Expects values between 0 and 1
+
+    def get_linguistic_features(self, input: str):
+        """
+        Linguistic analysis of the input sentence.
+        Returns a list of tokens, POS tags and lemmatised tokens.
+        """
+        tokens = []
+        pos_tags = []
+        lemmas = []
+        sentence = self.linguistic_pipeline(input)
+        for token in sentence:
+            tokens.append(token.text)
+            pos_tags.append(token.pos_)
+            lemmas.append(token.lemma)
+        return tokens, pos_tags, lemmas
+
+    def count_eligible_tokens(self, pos_tags):
+        """
+        Returns the number of words that will be replaced.
+        """
+        pos_count = 0
+        for pos in self.pos_to_change:
+            pos_count += pos_tags.count(pos)
+        return pos_count
+
+    def get_perturbation_candidates(self, input: str):
+        """
+        Samples tokens that will be replaced.
+        Generates and returns the top k contextual replacement candidates.
+        """
+        perturbation_dict = {}
+        tokens, pos_tags, lemmas = self.get_linguistic_features(input)
+        if self.count_eligible_tokens(pos_tags) == 0:
+            print(
+                "Warning: The sequence' ",
+                input,
+                " 'will remain unchanged as it didn't include the following POS tags:",
+                self.pos_to_change,
+            )
+            num_word_swaps = 0
+        else:
+            num_word_swaps = max(
+                1,
+                round(self.percentage * self.count_eligible_tokens(pos_tags)),
+            )
+        if self.verbose:
+            print(num_word_swaps, " tokens will be masked.\n")
+        random.seed(self.seed)
```

Comment:
It would be good to put this line to the __init__() function

Reply:
Done!